runtime/parachains/src/assigner_on_demand/mod.rs

```diff
@@ -0,0 +1,490 @@
+	) -> Result<FixedU128, SpotTrafficCalculationErr> {
+		// Return early if queue has no capacity.
+		if queue_capacity == 0 {
+			return Err(SpotTrafficCalculationErr::QueueCapacityIsZero)
+		}
+
+		// Return early if queue size is greater than capacity.
+		if queue_size > queue_capacity {
+			return Err(SpotTrafficCalculationErr::QueueSizeLargerThanCapacity)
+		}
+
+		// (queue_size / queue_capacity) - target_queue_utilisation
+		let queue_util_ratio = FixedU128::from_rational(queue_size.into(), queue_capacity.into());
+		let positive = queue_util_ratio >= target_queue_utilisation.into();
+		let queue_util_diff = queue_util_ratio.max(target_queue_utilisation.into()) -
+			queue_util_ratio.min(target_queue_utilisation.into());
+
+		// variability * queue_util_diff
+		let var_times_qud = queue_util_diff.saturating_mul(variability.into());
+
+		// variability^2 * queue_util_diff^2
+		let var_times_qud_pow = var_times_qud.saturating_mul(var_times_qud);
+
+		// (variability^2 * queue_util_diff^2)/2
+		let div_by_two: FixedU128;
+		match var_times_qud_pow.const_checked_div(2.into()) {
+			Some(dbt) => div_by_two = dbt,
+			None => return Err(SpotTrafficCalculationErr::Division),
+		}
+
+		// traffic * (1 + queue_util_diff) + div_by_two
+		if positive {
+			let new_traffic = queue_util_diff
+				.saturating_add(div_by_two)
+				.saturating_add(One::one())
+				.saturating_mul(traffic);
+			Ok(new_traffic.max(<T as Config>::TrafficDefaultValue::get()))
+		} else {
+			let new_traffic = queue_util_diff.saturating_sub(div_by_two).saturating_mul(traffic);
+			Ok(new_traffic.max(<T as Config>::TrafficDefaultValue::get()))
+		}
+	}
+
+	/// Adds an assignment to the on demand queue.
+	///
+	/// Paramenters:
+	/// - `assignment`: The on demand assignment to add to the queue.
+	/// - `location`: Whether to push this entry to the back or the front of the queue.
+	///               Pushing an entry to the front of the queue is only used when the scheduler
+	///               wants to push back an entry it has already popped.
+	pub fn add_parathread_assignment(
+		assignment: Assignment,
+		location: QueuePushDirection,
+	) -> Result<(), DispatchError> {
+		// Only parathreads are valid paraids for on the go parachains.
+		ensure!(<paras::Pallet<T>>::is_parathread(assignment.para_id), Error::<T>::InvalidParaId);
+
+		let config = <configuration::Pallet<T>>::config();
+
+		// Be explicit rather than implicit about the return value
+		let res = OnDemandQueue::<T>::try_mutate(|queue| {
+			// Abort transaction if queue is too large
+			ensure!(Self::queue_size() < config.on_demand_queue_max_size, Error::<T>::QueueFull);
+			match location {
+				QueuePushDirection::Back => queue.push_back(assignment),
+				QueuePushDirection::Front => queue.push_front(assignment),
+			};
+			Ok(())
+		});
+		res
+	}
+
+	/// Get the size of the on demand queue.
+	/// Returns:
+	/// - The size of the on demand queue, or failing to do so, the max size of the queue.
```

Comment:
This was a bit confusing. We are not actually failing in getting the size, this is just about the corner case that it does not fit into a u32. I would suggest to ensure via the interface and tests that this can never happen and then in the `Err` case we can still use the max size, but add a debug log, saying that an invariant was not upheld. 

Then we can simply remove the comment about failing in the doc, because it should never happen and if it does it is an internal bug and the caller should not need to bother with those.

Reply:
This is a private internal function so the caller is part of making sure the invariant is upheld, nevertheless let's do the debug log instead of the doc.